net/wire/src/network_address/onion_addr.rs

```diff
@@ -0,0 +1,224 @@
+    // Virtual port of the peer
+    pub port: u16,
+}
+
+/// Error enum at parsing onion addresses
+#[derive(Debug, Error)]
+pub enum OnionAddrParsingError {
+    #[error("Address is either too long or short, length: {0}")]
+    InvalidLength(usize),
+    #[error("Address contain non-utf8 code point at tld byte location: {0:x}")]
+    NonUtf8Char(u8),
+    #[error("This is not an onion address, Tld: {0}")]
+    InvalidTld(String),
+    #[error("Domain contains non base32 characters")]
+    NonBase32Char,
+    #[error("Invalid port specified")]
+    InvalidPort,
+}
+
+impl OnionAddr {
+    /// Attempt to create an [`OnionAddr`] from a complete .onion address string and a port.
+    ///
+    /// Return None if the supplied `addr` is invalid.
+    pub fn new(addr: &str, port: u16) -> Option<Self> {
+        Self::check_addr(addr).is_ok().then_some(Self {
+            domain: addr.as_bytes()[..56].try_into().ok()?,
+            port,
+        })
+    }
+
+    /// Establish if the .onion address is valid.
+    ///
+    /// Return `true` if valid, `false` otherwise.
+    pub fn check_addr(addr: &str) -> Result<(), OnionAddrParsingError> {
+        // v3 onion addresses are 62 characters long
+        if addr.len() != 62 {
+            return Err(OnionAddrParsingError::InvalidLength(addr.len()));
+        }
+
+        let Some((domain, tld)) = addr.split_at_checked(56) else {
+            return Err(OnionAddrParsingError::NonUtf8Char(addr.as_bytes()[56]));
+        };
+
+        // The ".onion" suffix must be located at the 57th byte.
+        if tld != ".onion" {
+            return Err(OnionAddrParsingError::InvalidTld(String::from(tld)));
+        }
+
+        // The domain part must only contain base32 characters.
+        if !domain
+            .as_bytes()
+            .iter()
+            .copied()
+            .all(|c| c.is_ascii_lowercase() || (b'2'..=b'7').contains(&c))
+        {
+            return Err(OnionAddrParsingError::NonBase32Char);
+        }
+
+        Ok(())
+    }
+
+    pub const fn port(&self) -> u16 {
+        self.port
+    }
+}
+
+/// Display for [`OnionAddr`] only print the **onion address**. Not the port.
+impl Display for OnionAddr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let str = self
+            .domain
+            .utf8_chunks()
+            .next()
+            .expect("Onion addresses are always containing UTF-8 characters.");
+
+        f.write_str(str.valid())?;
+        f.write_str(".onion")
+    }
+}
+
+/// [`OnionAddr`] parses an onion address **and a port**.
+impl FromStr for OnionAddr {
+    type Err = OnionAddrParsingError;
+
+    fn from_str(addr: &str) -> Result<Self, Self::Err> {
+        let (addr, port) = addr
+            .split_at_checked(62)
+            .ok_or(OnionAddrParsingError::InvalidLength(addr.len()))?;
+
+        // Port
+        let port: u16 = port
+            .starts_with(':')
+            .then_some(port[1..].parse().ok())
```

Comment:
this will panic if the string is empty, i.e has nothing after the `.onion`.

Reply:
I don't think you ran the same test:

```
byte index 1 is out of bounds of ``
thread 'network_address::onion_addr::tests::parse_invalid_onion_address_w_port' panicked at net/wire/src/network_address/onion_addr.rs:111:28:
byte index 1 is out of bounds of ``
```

You have fallen into a pretty common trap, `then_some` is a function that takes in a value `T` and outputs it if the `bool` is `true`. The value `T` is computed before the call and passed. What you want is something that takes in a closure/function and runs that closure/function to produce the value `T` if the bool is `Some`.

```suggestion
            .then(|| port[1..].parse().ok())
```